Sources/Utility/FSWatch.swift

```diff
@@ -0,0 +1,602 @@
+
+            let nfds = [fd, cancellationPipe[0]].reduce(0, max) + 1
+            // num fds, read fds, write fds, except fds, timeout
+            let selectRet = select(nfds, &rfds, nil, nil, nil)
+            // FIXME: Check for int signal.
+            assert(selectRet != -1)
+
+            // Return if we're cancelled.
+            if FD_ISSET(cancellationPipe[0], &rfds) {
+                return
+            }
+            assert(FD_ISSET(fd, &rfds))
+
+            let buf = UnsafeMutablePointer<CChar>.allocate(capacity: Inotify.eventSize)
+            // FIXME: We need to free the buffer.
+
+            let readLength = read(fd, buf, Inotify.eventSize)
+            // FIXME: Check for int signal.
+    
+            // Consume events.
+            var idx = 0
+            while idx < readLength {
+                let event = withUnsafePointer(to: &buf[idx]) {
+                    $0.withMemoryRebound(to: inotify_event.self, capacity: 1) {
+                        $0.pointee
+                    }
+                }
+
+                // Get the associated with the event.
+                var path = wds[event.wd]!
+
+                // FIXME: We need extract information from the event mask and
+                // create a data structure.
+                // FIXME: Do we need to detect and remove watch for directories
+                // that are deleted?
+
+                // Get the relative base name from the event if present.
+                if event.len > 0 {
+                    // Get the basename of the file that had the event.
+                    let basename = String(cString: buf + idx + MemoryLayout<inotify_event>.size)
+    
+                    // Construct the full path.
+                    // FIXME: We should report this path separately.
+                    path = path.appending(component: basename)
+                }
+
+                // Signal the reporter.
+                reportCondition.whileLocked {
+                    lastEventTime = Date()
+                    collectedEvents.append(path)
+                    reportCondition.signal()
+                }
+
+                idx += MemoryLayout<inotify_event>.size + Int(event.len)
+            }
+        }
+    }
+
+    /// Spawns a thread that collects events and reports them after the settle period.
+    private func startReportThread() {
+        let thread = Basic.Thread {
+            var endLoop = false
+            while !endLoop {
+
+                // Block until we timeout or get signalled.
+                self.reportCondition.whileLocked {
+                    var performReport = false
+
+                    // Block until timeout expires or wait forever until we get some event.
+                    if let lastEventTime = self.lastEventTime {
+                        let timeout = lastEventTime + Double(self.settle)
+                        let timeLimitReached = !self.reportCondition.wait(until: timeout)
+
+                        if timeLimitReached {
```

Comment:
Not 100% sure but it seems to me that all this dance with timeout events is to allow cancellation, is this right ?
If that's the case then I think you could avoid that and keep it simple by just closing the file descriptor when cancelling, as I mentioned in another line.

Reply:
No, this is for implementing the FSEventStream's "latency" feature, otherwise we end up sending a lot of callbacks. This collects the events until a settle period and then reports them at once.

Reply:
Ah ok, but I'd recommend to let the client control this and add an option for the client to setup how much latency it wants.